Add readme for kubernetes-overlord-extensions and update docs (#14674)
apache/druid: docs/development/extensions-contrib/k8s-jobs.md

## How it works

> The K8s extension builds a pod spec from each task using the specified pod adapter. All jobs are natively restorable, they are decoupled from the druid deployment, thus restarting pods or doing upgrades has no affect on tasks in flight.  They will continue to run and when the overlord comes back up it will start tracking them again.  

Reviewer: ```suggestion
The K8s extension builds a pod spec for each task using the specified pod adapter. All jobs are natively restorable, they are decoupled from the Druid deployment, thus restarting pods or doing upgrades has no affect on tasks in flight.  They will continue to run and when the overlord comes back up it will start tracking them again.  
```

## Configuration

To use this extension please make sure to  [include](../../configuration/extensions.md#loading-extensions)`druid-kubernetes-overlord-extensions` in the extensions load list for your overlord process.

> The extension uses `druid.indexer.runner.capacity` to limit the number of k8s jobs in flight. You should set this to a reasonable value. Additionally set the variable `druid.indexer.runner.namespace` to the namespace in which you are running druid.

Reviewer: > You should set this to a reasonable value.

Can we include some guidance on what is reasonable. I think mentioning that having this set to infinite means that the overlord will have more threads will be useful for an operator to tune their cluster.

> The extension uses `druid.indexer.runner.capacity` to limit the number of k8s jobs in flight. You should set this to a reasonable value. Additionally set the variable `druid.indexer.runner.namespace` to the namespace in which you are running druid.

Reviewer: ```suggestion
The extension uses `druid.indexer.runner.capacity` to limit the number of concurrent peons running in the Kubernetes cluster. Setting this value too high can cause the overlord to be OOMKilled as it spawns more threads to keep track of the peons. 
```

> Additionally set the variable `druid.indexer.runner.namespace` to the namespace in which you are running druid.

It feels like this belongs on the next line with the rest of the required configuration

All other configurations you had for the middle manager tasks must be moved under the overlord with one caveat, you must specify javaOpts as an array:
`druid.indexer.runner.javaOptsArray`, `druid.indexer.runner.javaOpts` is no longer supported.

If you are running without a middle manager you need to also use `druid.processing.intermediaryData.storage.type=deepstore`

### Custom Template Pod Adapter
The custom template pod adapter allows you to specify a pod template file per task type for more flexibility on how to define your pods. This adapter expects a [Pod Template](https://kubernetes.io/docs/concepts/workloads/pods/#pod-templates) to be available on the overlord's file system. This pod template is used as the base of the pod spec for the Kubernetes Job. You can override things like labels, environment variables, resources, annotation, or even the base image with this template. To enable this pod adapter you can specify the runtime property `druid.indexer.runner.k8s.adapter.type: customTemplateAdapter`

The base pod template must be specified as the runtime property `druid.indexer.runner.k8s.podTemplate.base: /path/to/basePodSpec.yaml`

> Task specific pod templates must be specified as the runtime property `druid.indexer.runner.k8s.podTemplate.{taskType}: /path/to/taskSpecificPodSpec.yaml` where {taskType} is the name of the task type i.e `index_parallel`

Reviewer: ```suggestion
Task specific pod templates can be specified as the runtime property `druid.indexer.runner.k8s.podTemplate.{taskType}: /path/to/taskSpecificPodSpec.yaml` where {taskType} is the name of the task type i.e `index_parallel`
```

The following is an example Pod Template that uses the regular druid docker image.
```
apiVersion: "v1"
kind: "PodTemplate"
template:
  metadata:
    annotations:

>       custom-annotation: "hello"

Reviewer: Can you provide a real world example of an annotation that would be helpful in the template. Same comment about the labels.

    labels:
      custom-label: "hello"
  spec:
    affinity: {}
    containers:
    - command:
        - sh
        - -c
        - |
          /peon.sh /druid/data 1
      env:
      - name: CUSTOM_ENV_VARIABLE
        value: "hello"

>       image: apache/druid:26.0.0

Reviewer: ```suggestion
      image: apache/druid:{{DRUIDVERSION}}
```

There is a script that runs and replaces {{DRUIDVERSION}} with the version of the docs where this is published

      name: main
      ports:
      - containerPort: 8091
        name: druid-tls-port
        protocol: TCP
      - containerPort: 8100
        name: druid-port
        protocol: TCP
      resources:
        limits:

>           cpu: "1"

Reviewer: Do we need to set a limit?

Author: that depends on what you are doing but usually it's best practice to do so

Reviewer: https://home.robusta.dev/blog/stop-using-cpu-limits - I read this blog post that seems to advise against setting CPU limits

Author: i would say my intention with this template is to show all the possible configurations you can set on the pod template, it's easier to see the configuration in docs and look up what it does and remove it if not needed than to try to figure out if a property you found somewhere else can be set

it's not necessarily intended to be a recommended configuration

          memory: 2400M
        requests:
          cpu: "1"

>           memory: 2400M

Reviewer: This requests seems high given that Xmx is 1G below. Does it need to be this high?

Author: there was also 1G of direct mem so this is a pretty reasonable request i think

Reviewer: oops, missed the 1G of direct memory

      volumeMounts:

>       - mountPath: /opt/druid/conf/druid/cluster/master/coordinator-overlord # runtime props are still mounted in this location because that's where peon.sh looks for configs

Reviewer: This seems a little confusing. Should `peon.sh` to look for runtime properties in another location for the peons?

Author: i think it would probably make sense, i can do this in a separate pr

        name: nodetype-config-volume
        readOnly: true
      - mountPath: /druid/data
        name: data-volume
      - mountPath: /druid/deepstorage
        name: deepstorage-volume
    restartPolicy: "Never"
    securityContext:
      fsGroup: 1000
      runAsGroup: 1000

>       runAsUser: 1000

Reviewer: Can you explain what these settings are please

Author: this is the user the druid process runs on. i think we specify 1000 here because that user is created in the druid docker image that i used in this example.

metadata:
    name: druid-tiny-cluster-peons-config
    namespace: default
apiVersion: v1
data:
    jvm.config: |-
        -server
        -XX:MaxDirectMemorySize=1000M
        -Duser.timezone=UTC
        -Dfile.encoding=UTF-8
        -Dlog4j.debug
        -Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager
        -Djava.io.tmpdir=/druid/data
        -Xmx1024M
        -Xms1024M
    log4j2.xml: |-
        <?xml version="1.0" encoding="UTF-8" ?>
        <Configuration status="WARN">
            <Appenders>
                <Console name="Console" target="SYSTEM_OUT">
                    <PatternLayout pattern="%d{ISO8601} %p [%t] %c - %m%n"/>
                </Console>
            </Appenders>
            <Loggers>
                <Root level="info">
                    <AppenderRef ref="Console"/>
                </Root>
            </Loggers>
        </Configuration>
    runtime.properties: |
        druid.port=8100
        druid.service=druid/peon
        druid.server.http.numThreads=5
        druid.indexer.task.baseTaskDir=/druid/data
        druid.indexer.runner.type=k8s
        druid.peon.mode=remote
        druid.indexer.task.encapsulatedTask=true
```

> 
> Additional Configuration

Reviewer: ```suggestion
```

|`druid.indexer.runner.annotations`| `JsonObject`    | Additional annotations you want to add to peon pod                                                                                                                                                                                               |`{}`|No|
|`druid.indexer.runner.peonMonitors`| `JsonArray`     | Overrides `druid.monitoring.monitors`. Use this property if you don't want to inherit monitors from the Overlord.                                                                                                                                |`[]`|No|
|`druid.indexer.runner.graceTerminationPeriodSeconds`| `Long`          | Number of seconds you want to wait after a sigterm for container lifecycle hooks to complete.  Keep at a smaller value if you want tasks to hold locks for shorter periods.                                                                      |`PT30S` (K8s default)|No|

> |`druid.indexer.runner.capacity`| `Integer`       | Number of concurrent jobs that can be sent to Kubernetes.                                                                                                                                                                                        |`2147483647 `|No|

Reviewer: ```suggestion
|`druid.indexer.runner.capacity`| `Integer`       | Number of concurrent jobs that can be sent to Kubernetes.                                                                                                                                                                                        |`2147483647`|No|
```
